docs/sequence-improvement-design.md

```diff
@@ -0,0 +1,123 @@
+between `lsft` being applied to only `a` or to both `a` and `b`.
+
+The feature will codenamed "seqchords" for now.
+
+## An exploration of an idea: track releases?
+
+Today, the sequence `(lsft a b c)` doesn't care when the `lsft`,
+or even `a` or `b` are released relative to when the subsequent keys
+are pressed. However, with seqchords, the code could be potentially changed to
+make sequences release-aware.
+
+It seems a little difficult to integrate
+this into the trie structure used to track sequences though.
+With an implementation that is release-aware,
+it seems like the code would need to figure out how to conditionally
+add release events to the trie, depending if the seq was
+`(lsft a)` or `(S-a)`
+
+For now, I think a different approach would be better.
+
+## A different idea: modify presses held with mod keys.
+
+The current sequence type is `Vec<u16>` since keys don't fit into `u8`.
+However, there are fewer than 1024 (2^10) keys total.
+That means there are 6 bits to play with.
+6 bits are enough for the types of modifiers (of which there are 4),
+but differentiating both sides (increases to 8).
+Perhaps one only cares to use both left and right shifts though, and maybe
+both left and right alts.
+One could also use a `u32` instead, but that seems unnecessary for now.
+I see no backwards compatibility issues if one
+desired that change in the future.
+
+With this in mind, while modifiers are held, set the upper unused bits of
+the stored `u16` values.
+
+### Backwards compatibility?
+
+This does mean that `(lsft a b)` behaves differently
+with vs. without seqchords.
+Unless maybe the code automatically generates the various permutations
+of this type of sequence, but that seems complicated.
+Or maybe have a `u16` with a special bit pattern that could be used
+to differentiate between `(S-(a b))` and `(lsft a b)`.
+For now, let's say that the bit pattern is `0xFFFF`.
+If a modifier is pressed and the the sequence `[..., <mod>, 0xFFFF]`
+exists in the trie: continue processing the sequence in mod-aware mode.
+
+OR for simplicity, just say "screw backwards compatibility" and force users
+to be clear about what they mean and define the extra permutations, if they
+want them. I prefer this.
+
+### Data format examples
+
+Let's begin the description of the new data format.
+Since shifted keys seem like they will be the main use case for seqchords,
+only that will be described in this document for now.
+Here are the numerical key values relevant to the examples.
+
+- `a:    0x001E`
+- `b:    0x0030`.
+- `lsft: 0x002A`.
+
+This differs by OS, but that's not important.
+
+The transformation of `(lsft a b)` to a sequence in the trie today
+looks like:
+
+- `[0x002A, 0x001E, 0x0030]`
+
+This will remain unchanged with seqchords.
+Let's say that chorded keys using `lsft`
+will have the otherwise-unused MSB (bit 15) set.
+
+The transformation of some sequences using chords will be:
+
+1. `(S-(a b)) => [0x802A, 0x801E, 0x8030]`
```

Comment:
Shouldn't this transform to `[0x802A, 0x801E, 0x802A, 0x8030]`? Though I don't know the internals of sequences at all, so I might not be understanding something.

Reply:
The way sequences work is that a new `u16` is added to the in-progress sequence when a key is newly pressed, and that is checked against the trie.

In this case you're describing point 3: `S-a S-b`.

The two cases `S-(a b)` and `S-a S-b` are not transformed identically. The first means "press shift once and hold throughout" whereas the second means that shift is pressed twice.